src/libradosstriper/RadosStriperImpl.cc

```diff
@@ -1188,7 +1188,6 @@ int libradosstriper::RadosStriperImpl::aio_write_in_open_object(const std::strin
   // we need 3 references as striper_write_aio_req_complete will release two and
   // striper_write_aio_req_safe will release one
   auto cdata = ceph::make_ref<WriteCompletionData>(this, soid, lockCookie, c);
-  m_ioCtxImpl->get();
```

Comment:
I think instead of avoid the bump up ref_cnt we should bump down ref_cnt either in WriteCompletionData destructor or during destroying of c(AioCompletionImpl) object, it happens while call comp->release(); by the application. During AioCompletionImp::release we can bump down ref_cnt.

Reply:
The execution of `m_ioCtxImpl->get()` is a specific behavior of the `RadosStriperImpl::aio_write_in_open_object` function, whereas `AioCompletionImp::release()` is a general destruction operation. It is not appropriate to decrease the reference count here because other asynchronous interfaces do not increase the reference count.

Reply:
In that case you can do during destruction of WriteCompletionData object.

Reply:
Can we use a hook function? @mohit84 
```c++
--- a/src/libradosstriper/RadosStriperImpl.cc
+++ b/src/libradosstriper/RadosStriperImpl.cc
@@ -21,6 +21,7 @@
 #include <sstream>
 #include <iomanip>
 #include <algorithm>
+#include <functional>
 
 #include "include/types.h"
 #include "include/uuid.h"
@@ -245,12 +246,14 @@ void ReadCompletionData::complete_unlock(int r) {
  * function in asynchronous write operations
  */
 struct WriteCompletionData : CompletionData {
+  using CompleteHook = std::function<void()>;
   /// safe completion handler
   librados::IoCtxImpl::C_aio_Complete *m_safe;
   /// completion object for the unlocking of the striped object at the end of the write
   librados::AioCompletion *m_unlockCompletion;
   /// return code of write completion, to be remembered until unlocking happened
   int m_writeRc;
+  CompleteHook complete_hook;
   /// constructor
   WriteCompletionData(libradosstriper::RadosStriperImpl * striper,
                      const std::string& soid,
@@ -265,6 +268,8 @@ struct WriteCompletionData : CompletionData {
   void complete_unlock(int r);
   /// safe method
   void safe(int r);
+  /// complete hook
+  void set_complete_hook(CompleteHook hook);
 };
 
 WriteCompletionData::WriteCompletionData
@@ -282,7 +287,12 @@ WriteCompletionData::WriteCompletionData
 
 WriteCompletionData::~WriteCompletionData() {
   m_unlockCompletion->release();
   if (m_safe) delete m_safe;
+  if (complete_hook) {
+      complete_hook();
+  }
 }

+void WriteCompletionData::set_complete_hook(CompleteHook hook) {
+    complete_hook = hook
+}

int libradosstriper::RadosStriperImpl::aio_write_in_open_object(const std::strin
   cdata->get(); // local ref
   m_ioCtxImpl->get();
   c->io = m_ioCtxImpl;
+  cdata->set_complete_hook([this](){
+    m_ioCtxImpl->put();
+  });
  ...
```

Reply:
We can use this or we can make it simple also something like below
```
-- a/src/libradosstriper/RadosStriperImpl.cc
+++ b/src/libradosstriper/RadosStriperImpl.cc
@@ -284,7 +284,11 @@ WriteCompletionData::WriteCompletionData
 
 WriteCompletionData::~WriteCompletionData() {
   m_unlockCompletion->release();
-  if (m_safe) delete m_safe;
+  if (m_safe) {
+    if (m_safe->c->io)
+      m_safe->c->io->put();
+    delete m_safe;
+  }
 }
 

```